soroban-env-host/src/host.rs

```diff
@@ -799,9 +818,59 @@ impl Host {
-                        .ok_or_else(|| self.err_general("function not found"))
+                let frame = TestContractFrame::new(id.clone(), func.clone());
+                let panic = frame.panic.clone();
+                return self.with_frame(Frame::TestContract(frame), || {
+                    use std::any::Any;
+                    use std::panic::{catch_unwind, set_hook, take_hook, AssertUnwindSafe};
+                    type PanicVal = Box<dyn Any + Send>;
+
+                    // We're directly invoking a native rust contract here,
+                    // which we allow only in local testing scenarios, and we
+                    // want it to behave as close to the way it would behave if
+                    // the contract were actually compiled to WASM and running
+                    // in a VM.
+                    //
+                    // In particular: if the contract function panics, if it
+                    // were WASM it would cause the VM to trap, so we do
+                    // something "as similar as we can" in the native case here,
+                    // catch the native panic and attempt to continue by
+                    // translating the panic back to an error, so that
+                    // `with_frame` will rollback the host to its pre-call state
+                    // (as best it can) and propagate the error to its caller
+                    // (which might be another contract doing try_call).
+                    //
+                    // This is somewhat best-effort, but it's compiled-out when
+                    // building a host for production use, so we're willing to
+                    // be a bit forgiving.
+                    let closure = AssertUnwindSafe(move || cfs.call(&func, self, args));
+                    let existing_panic_hook = take_hook();
+                    set_hook(Box::new(|_| ()));
+                    let res: Result<Option<RawVal>, PanicVal> = catch_unwind(closure);
+                    set_hook(existing_panic_hook);
+                    match res {
+                        Ok(Some(rawval)) => Ok(rawval),
+                        Ok(None) => Err(self.err(
+                            DebugError::general()
+                                .msg("error '{}': calling unknown contract function '{}'")
+                                .arg::<RawVal>(func.into()),
+                        )),
+                        Err(_) => {
+                            // Return an error indicating the contract function
+                            // panicked. If if was a panic generated by a
+                            // CheckedEnv-upgraded HostError, it had its status
+                            // captured by VmCallerCheckedEnv::escalate_error_to_panic:
+                            // fish the Status stored in the frame back out and
+                            // propagate it.
+                            let msg = "caught panic with '{}' in contract function '{}'";
```

Comment:
❔ It appears this will make the debug events significantly different when running in WASM VM vs native. I'm not sure that's a good thing. Do we need to create the debug event for the escalation and panic caught? I'm not attached to them not being present, I'm just wondering if it will cause confusion.

Reply:
I can make them _more similar_ if you like (eg. remove the escalation event, rephrase the "caught" message to be somewhat closer to the "VM trapped with host error" error we make if there is a VM trap with host error) but .. I'm a little hesitant here. I would also say:

1. Debug events are for _debugging_, not something users should rely on the content of. We want to retain freedom to change debug events fairly freely over time, and we're probably going to add quite a lot. It may well be that we will be emitting debug events on VM setup/teardown or host function calls -- this will necessarily make these two code paths emit different debug events _intentionally_, because different things are happening. 
2. VM-trap vs. panic-and-catch are fundamentally slightly-different mechanisms, and they're never going to be identical. For example an out-of-bounds memory read or divide-by-zero will deliver a signal to the process rather than panic, whereas a VM will trap them. And even more concretely: the budgeting of native contracts is fundamentally different, and panics arising from out-of-budget will always be different.

Reply:
I think removing the additional escalation event might be a good idea, but keeping the text for the other one would be fine. I agree calling it a VM trap would be misleading and we shouldn't say anything untrue. I agree people shouldn't depend on debug events and they should change, but if the debug events make the code and behavior appear significantly different that may be confusing.